### PR TITLE
Added Chat client package

### PR DESCRIPTION
**Pull request by volodkevych**
## Problem
Created basic Chat client package with MynahUI in it. All the needed commands will be implemented in the following CRs. This CR implements only few events/commands, mainly as an example:
- `send to prompt`
- `focus`
- `tabs added/removed/updated`
- `ui ready`

Types for all the events/commands will be moved to separate package in following CRs.

`vscode` client was modified to load Chat client and trigger events/commands - mainly for testing purposes.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,7 +23,7 @@
         "test-integ": "npm run compile && npm run test-integ --workspaces --if-present",
         "test-unit": "npm run compile && npm run test-unit --workspaces --if-present",
         "update-runtimes": "cp ../language-server-runtimes/aws-*.tgz bin/ && npm update @aws/language-server-runtimes",
-        "package": "npm install && npm run compile && npm run package-x64 --workspaces --if-present"
+        "package": "npm install && npm run compile && npm run package-x64 --workspaces --if-present && npm run package --workspaces --if-present"
```
Comment (rahmaniaam) [inline package.json:26]: nit: we should refactor all the packages to just use `npm run package`, just makes the script too long here

File: chat-client/src/client/chat.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    UI_FOCUS,
+    UI_IS_READY,
+    UiMessage,
+} from '../contracts/uiContracts'
+import { Messager, OutboundChatApi } from './messager'
+import { InboundChatApi, createMynahUI } from './mynahUi'
+import { TabFactory } from './tabs/tabFactory'
+import { TabStorage } from './tabs/tabStorage'
+
+export const createChat = (clientApi: { postMessage: (msg: UiMessage | ServerMessage) => void }) => {
```
Comment (rahmaniaam) [inline chat-client/src/client/chat.ts:20]: Not limited to this file, but can we add unit tests? does it make sense to have any for the chat client?

**Review by rahmaniaam — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (volodkevych) [reply package.json:26]: Maybe we should: I did not see any usage of `package` & `package-x64` in one package. But it's a very low priority. But it' a low priority.

File: chat-client/src/client/chat.ts (hunk 2 above)
Comment (volodkevych) [reply chat-client/src/client/chat.ts:20]: Yes, will do it as a followup.

File: chat-client/src/client/chat.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                command: UI_IS_READY,
+            })
+
+            window.addEventListener('message', handleMessage)
+            window.addEventListener('focus', handleApplicationFocus)
+            window.addEventListener('blur', handleApplicationFocus)
+        },
+    }
+
+    const connector = new Messager(chatApi)
```
Comment (viktorsaws) [inline chat-client/src/client/chat.ts:78]: super minor nit: naming could be made consistent, everywhere else this is referred to as `messager`

File: chat-client/src/client/chat.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { InboundChatApi, createMynahUI } from './mynahUi'
+import { TabFactory } from './tabs/tabFactory'
+import { TabStorage } from './tabs/tabStorage'
+
+export const createChat = (clientApi: { postMessage: (msg: UiMessage | ServerMessage) => void }) => {
+    // eslint-disable-next-line semi
+    let mynahApi: InboundChatApi
+
+    const sendMessageToClient = (message: UiMessage | ServerMessage) => {
+        clientApi.postMessage(message)
```
Comment (viktorsaws) [inline chat-client/src/client/chat.ts:25]: What is referred to as `client` here?
If I read this correctly (correcty me if I'm wrong), this is connecting MynahUI Chat Client inside of webview with outside proxy over postMessage in the hosting application. If that's true, can this be renamed to something more explicit, `sendMessageToHost` perhaps?

I found it a bit confusing to understand the direction of messages flow, perhaps changing the name or adding some documentation could help here.

Comment (volodkevych) [reply chat-client/src/client/chat.ts:25]: `clientApi` is `vscode` api if the extension is `vscode` extension. `postMessage` cannot be renamed to `sendMessageToHost`, `clientApi` can be renamed to whatever makes it more clear.

File: chat-client/src/client/tabs/tabStorage.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    private tabs: Map<string, Tab> = new Map()
+    private lastSelectedTab?: Tab
+
+    public addTab(tab: Tab) {
+        if (this.tabs.has(tab.id)) {
+            return
+        }
+        this.tabs.set(tab.id, tab)
+        if (tab.isSelected) {
+            this.setSelectedTab(tab.id)
```
[the remark this answers is not in the document]
Comment (volodkevych) [reply chat-client/src/client/tabs/tabStorage.ts:16]: If tab with the same ID already exists, the request to add a tab (either selected or not selected) is ignored. This is the logic in current MynahUI. If we need to change it in the future, we will change it.

File: chat-client/src/client/mynahUi.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+        messager.onSendToPrompt(params, tabId)
+    }
+
+    return {
+        sendToPrompt: sendToPrompt,
+    }
+}
+
+const uiComponentsTexts = {
```
[the remark this answers is not in the document]
Comment (volodkevych) [reply chat-client/src/client/mynahUi.ts:84]: Not sure at this point of time. If we need, we'll change that.

File: chat-client/package.json [hunk 7]
```diff
@@ -0,0 +1,37 @@
+{
+    "name": "@aws/q-chat-client",
```
[the remark this answers is not in the document]
Comment (volodkevych) [reply chat-client/package.json:2]: I can rename if there are suggestions.

**Pull request merged**